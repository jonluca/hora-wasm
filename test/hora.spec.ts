import { describe, expect, it } from "vitest";
import { getHora, HoraJsType } from "../index.ts";
import seedrandom from "seedrandom";
import {
  BruteForceIndexUsize,
  HNSWIndexUsize,
  IVFPQIndexUsize,
  PQIndexUsize,
  SSGIndexUsize,
} from "../pkg/horajs.js";

const DIMENSION = 50;
describe("hora-wasm", async () => {
  let hora: HoraJsType = await getHora();
  await hora.init_env();

  const SEARCH_INDEXES = [
    "euclidean",
    "manhattan",
    "dot_product",
    "cosine_similarity",
    "angular",
  ];

  for (const searchIndex of SEARCH_INDEXES) {
    // recreate index for every search type
    const INDEXES = [
      hora.BruteForceIndexUsize.new(DIMENSION),
      hora.HNSWIndexUsize.new(DIMENSION, 1000000, 32, 64, 20, 500, false),
      // TODO - these indexes don't work currently
      // hora.IVFPQIndexUsize.new(DIMENSION, 1000000, 32, 64, 20, 500),
      // hora.PQIndexUsize.new(DIMENSION, 1000000, 32, 64),
      // hora.SSGIndexUsize.new(DIMENSION, 1000000, 32, 64, 20, 500),
    ];

    for (let i = 0; i < INDEXES.length; i++) {
      const idx = INDEXES[i];

      let name = idx.name();
      const rng = seedrandom(`horajs-consistent-rng-${name}-${searchIndex}`);

      const getFeatures = (count: number) => {
        const features: number[][] = [];
        for (let i = 0; i < count; i++) {
          const feature: number[] = [];
          for (let j = 0; j < DIMENSION; j++) {
            feature.push(rng());
          }
          features.push(feature);
        }
        return features;
      };

      it(`${name} - creates an index and searches over it using ${searchIndex}`, () => {
        const features = getFeatures(1000);
        for (let i = 0; i < features.length; i++) {
          const feature = features[i];
          idx.add(Float32Array.from(feature), i); // add point
        }
        idx.build(searchIndex); // build index
        const featureToSearch = getFeatures(1)[0];
        const search_result = idx.search(
          Float32Array.from(featureToSearch),
          10
        );
        expect(search_result).toBeInstanceOf(Uint32Array);
        expect(search_result.length).toBe(10);
        expect(search_result).toMatchSnapshot();
      });
      it(`${name} - serializes indexes properly`, () => {
        const features = getFeatures(1000);
        for (let i = 0; i < features.length; i++) {
          const feature = features[i];
          idx.add(Float32Array.from(feature), i); // add point
        }
        idx.build(searchIndex); // build index
        const featureToSearch = getFeatures(1)[0];
        let searchVector = Float32Array.from(featureToSearch);
        const search_result = idx.search(searchVector, 10);
        const serialized = idx.dump_index();
        const className = idx.constructor.name;
        const reconstructedIdx = (
          hora[className as keyof typeof hora] as any
        ).load_index(serialized);
        const new_search_result = reconstructedIdx.search(searchVector, 10);

        expect(search_result).toEqual(new_search_result);
      });
    }
  }
});
