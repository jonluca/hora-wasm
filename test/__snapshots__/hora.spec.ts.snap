// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`hora-wasm > BruteForceIndex - bulk inserts and indexes using angular 1`] = `
Uint32Array [
  1557,
  485,
  2325,
  108,
  6752,
  5914,
  629,
  7498,
  6807,
  3901,
]
`;

exports[`hora-wasm > BruteForceIndex - bulk inserts and indexes using cosine_similarity 1`] = `
Uint32Array [
  6181,
  3773,
  1604,
  7526,
  5533,
  4073,
  6743,
  7005,
  9119,
  7753,
]
`;

exports[`hora-wasm > BruteForceIndex - bulk inserts and indexes using dot_product 1`] = `
Uint32Array [
  5711,
  9906,
  2240,
  9203,
  3739,
  407,
  438,
  849,
  9330,
  5168,
]
`;

exports[`hora-wasm > BruteForceIndex - bulk inserts and indexes using euclidean 1`] = `
Uint32Array [
  6630,
  9398,
  3772,
  844,
  2871,
  7665,
  4808,
  2736,
  2964,
  3982,
]
`;

exports[`hora-wasm > BruteForceIndex - bulk inserts and indexes using manhattan 1`] = `
Uint32Array [
  8003,
  5652,
  2481,
  9385,
  3794,
  7833,
  4955,
  5795,
  8458,
  6406,
]
`;

exports[`hora-wasm > BruteForceIndex - creates an index and searches over it using angular 1`] = `
Uint32Array [
  3543,
  6923,
  9810,
  3697,
  5139,
  4667,
  4828,
  842,
  9707,
  780,
]
`;

exports[`hora-wasm > BruteForceIndex - creates an index and searches over it using cosine_similarity 1`] = `
Uint32Array [
  954,
  4705,
  8498,
  1305,
  8684,
  1805,
  5696,
  5659,
  4128,
  16,
]
`;

exports[`hora-wasm > BruteForceIndex - creates an index and searches over it using dot_product 1`] = `
Uint32Array [
  9906,
  9203,
  4745,
  2240,
  8882,
  9190,
  1686,
  2583,
  1525,
  8902,
]
`;

exports[`hora-wasm > BruteForceIndex - creates an index and searches over it using euclidean 1`] = `
Uint32Array [
  4447,
  9942,
  1297,
  4559,
  6229,
  9135,
  8416,
  3702,
  2547,
  1274,
]
`;

exports[`hora-wasm > BruteForceIndex - creates an index and searches over it using manhattan 1`] = `
Uint32Array [
  2500,
  7350,
  3298,
  1115,
  374,
  1570,
  7357,
  7654,
  5979,
  9176,
]
`;

exports[`hora-wasm > HNSWIndex - bulk inserts and indexes using angular 1`] = `
Uint32Array [
  1457,
  7695,
  9455,
  6456,
  6563,
  2112,
  9611,
  1050,
  7617,
  3853,
]
`;

exports[`hora-wasm > HNSWIndex - bulk inserts and indexes using cosine_similarity 1`] = `
Uint32Array [
  4884,
  4295,
  8965,
  9473,
  6210,
  7921,
  8963,
  3177,
  7933,
  9477,
]
`;

exports[`hora-wasm > HNSWIndex - bulk inserts and indexes using dot_product 1`] = `
Uint32Array [
  7114,
  345,
  6075,
  8341,
  3243,
  116,
  7603,
  5763,
  3837,
  12,
]
`;

exports[`hora-wasm > HNSWIndex - bulk inserts and indexes using euclidean 1`] = `
Uint32Array [
  9816,
  8672,
  3002,
  3637,
  8658,
  6182,
  1631,
  5233,
  6512,
  4560,
]
`;

exports[`hora-wasm > HNSWIndex - bulk inserts and indexes using manhattan 1`] = `
Uint32Array [
  1810,
  4831,
  2087,
  3667,
  8800,
  7215,
  7862,
  5523,
  9936,
  8319,
]
`;

exports[`hora-wasm > HNSWIndex - creates an index and searches over it using angular 1`] = `
Uint32Array [
  1971,
  1675,
  6174,
  889,
  3077,
  6913,
  6083,
  6680,
  8151,
  7249,
]
`;

exports[`hora-wasm > HNSWIndex - creates an index and searches over it using cosine_similarity 1`] = `
Uint32Array [
  8714,
  6266,
  1211,
  8139,
  5770,
  4833,
  3954,
  5105,
  7049,
  9883,
]
`;

exports[`hora-wasm > HNSWIndex - creates an index and searches over it using dot_product 1`] = `
Uint32Array [
  8227,
  1490,
  3686,
  552,
  116,
  9373,
  4115,
  1313,
  8058,
  7648,
]
`;

exports[`hora-wasm > HNSWIndex - creates an index and searches over it using euclidean 1`] = `
Uint32Array [
  7711,
  7956,
  6520,
  4881,
  8254,
  1706,
  8553,
  3615,
  3086,
  9205,
]
`;

exports[`hora-wasm > HNSWIndex - creates an index and searches over it using manhattan 1`] = `
Uint32Array [
  8243,
  9458,
  3108,
  305,
  6057,
  3332,
  2907,
  8454,
  7082,
  3232,
]
`;
